File: user-service/rabbitmq/consumer.go
package rabbitmq

import (
	"log"

	amqp "github.com/rabbitmq/amqp091-go"
)

type RabbitMQConsumer struct {
	connectionManager *RabbitMQConnectionManager
	callback          func(amqp.Delivery)
	exchange          string
	routingKeys	   []string
	channel           *amqp.Channel
}

// NewRabbitMQConsumer creates a new RabbitMQConsumer instance
func NewRabbitMQConsumer(connectionManager *RabbitMQConnectionManager, callback func(amqp.Delivery), exchange string, routingKeys []string) *RabbitMQConsumer {
	return &RabbitMQConsumer{
		connectionManager: connectionManager,
		callback:          callback,
		exchange:          exchange,
		routingKeys:       routingKeys,
	}
}

// Consume establishes the channel, declares the exchange, binds the queue, and starts consuming messages
func (r *RabbitMQConsumer) Consume() {
	var err error
	r.channel = r.connectionManager.GetChannel()

	// Declare exchange
	err = r.channel.ExchangeDeclare(
		r.exchange,   // exchange name
		"topic",     // exchange type
		true,         // durable
		false,        // auto-deleted
		false,        // internal
		false,        // no-wait
		nil,          // arguments
	)
	if err != nil {
		log.Fatalf("Failed to declare exchange: %v", err)
	}

	// Declare exclusive queue
	queue, err := r.channel.QueueDeclare(
		"",    // name (empty for exclusive, generated name)
		false, // durable
		false, // delete when unused
		true,  // exclusive
		false, // no-wait
		nil,   // arguments
	)
	if err != nil {
		log.Fatalf("Failed to declare queue: %v", err)
	}

	for _, routingKey := range r.routingKeys {
		err = r.channel.QueueBind(
			queue.Name,   // queue name
			routingKey,   // routing key
			r.exchange,   // exchange name
			false,        // no-wait
			nil,          // arguments
		)
		if err != nil {
			log.Fatalf("Failed to bind queue: %v", err)
		}
	}
	
	if err != nil {
		log.Fatalf("Failed to bind queue: %v", err)
	}

	// Start consuming messages
	msgs, err := r.channel.Consume(
		queue.Name, // queue name
		"",         // consumer tag
		true,       // auto-ack
		false,      // exclusive
		false,      // no-local
		false,      // no-wait
		nil,        // args
	)
	if err != nil {
		log.Fatalf("Failed to start consuming messages: %v", err)
	}

	// Handle incoming messages
	for msg := range msgs {
		r.callback(msg)
	}
}

// StartConsuming starts the consumer in a separate goroutine
func (r *RabbitMQConsumer) StartConsuming() {
	log.Printf("Starting consumer for exchange: %s", r.exchange)
	go r.Consume() // Run Consume in a separate goroutine
}

File: activity-service/app/consumer.py
import threading
import logging
from connection_manager import RabbitMQConnectionManager

logger = logging.getLogger(__name__)


class RabbitMQConsumer:
    def __init__(self, connection_manager: RabbitMQConnectionManager,callback, exchange, routing_keys: list[str], queue_name=""):
        self.connection_manager = connection_manager
        self.callback = callback
        self.exchange = exchange
        self.routing_keys = routing_keys
        self.queue_name = queue_name
        self.channel = None

    def consume(self):
        self.channel = self.connection_manager.get_channel()
        self.channel.exchange_declare(exchange=self.exchange, exchange_type='topic', durable=True)
        result = self.channel.queue_declare(queue=self.queue_name, exclusive=True)
        queue_name = result.method.queue
        for routing_key in self.routing_keys:
            self.channel.queue_bind(exchange=self.exchange, queue=queue_name, routing_key=routing_key)
        self.channel.basic_consume(queue=queue_name, on_message_callback=self.callback, auto_ack=True)
        self.channel.start_consuming()
    
    def start_consuming(self):
        logger.info(f"Starting consumer thread for exchange:{self.exchange}")
        thread = threading.Thread(target=self.consume, daemon=True)
        thread.start()


